Created wrapper + tests for the Kademlia Library

thompalex opened the pull request:
Modified the peer class by creating a callback function for rpc_store in the Kademlia library to access the key and value.

NewServer overrides the _create_protocol method from Kademlia's network.Server to create an instance of our protocol.

Added tests for our NewServer and NewProtocol -> 100% testing coverage of this code.

zimmermatt commented on src/main/py/peer/peer.py:
```diff
@@ -80,16 +82,40 @@ def commission_art_piece(self) -> None:
+        self.logger.info("Data stored with key: %s", key)
+        self.logger.info("Data stored with value: %s", value)
+        artwork_object = pickle.loads(value)
+        if isinstance(artwork_object, Artwork):
```
Might be useful to have an `else` that logs that a non-artwork instance, if that's unexpected.

zimmermatt commented on src/main/py/server/network.py:
```diff
@@ -0,0 +1,46 @@
+from server.protocol import NewProtocol
+
+
+class NewServer(Server):
```
Let's brainstorm some more descriptive names :)

zimmermatt commented on src/main/py/server/protocol.py:
```diff
@@ -0,0 +1,36 @@
+from kademlia.protocol import KademliaProtocol
+
+
+class NewProtocol(KademliaProtocol):
```
Let's brainstorm some more descriptive names :)

zimmermatt commented on src/main/py/server/protocol.py:
```diff
@@ -0,0 +1,36 @@
+            ksize (int): The k parameter for Kademlia.
+            callback_function (function): The callback function to be called on rpc_store
+        """
+        self.callback_function = callback_function
```
The name for the callback in `NewServer` is better IMO.

zimmermatt commented on src/test/py/server/new_server_test.py:
```diff
@@ -0,0 +1,52 @@
+        """Test the store method of NewServer"""
+        await self.node1.listen(8468)
+        await self.node2.listen(8469)
+        await self.node2.bootstrap([("127.0.0.1", 8468)])
```
Let's see how it goes, but spinning up/tearing down servers may be a source of instability in the test suite.

zimmermatt commented on src/test/py/server/new_server_test.py:
```diff
@@ -0,0 +1,52 @@
+    async def test_store(self):
+        """Test the store method of NewServer"""
+        await self.node1.listen(8468)
+        await self.node2.listen(8469)
```
Please use ports in the dynamic range: 49152 - 65535.

See: https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt

zimmermatt commented on src/main/py/peer/peer.py:
```diff
@@ -47,35 +47,57 @@ async def send_deadline_reached(self, commission: Artwork) -> None:
+                self.logger.error("Commission failed to send")
+        except TypeError:
+            self.logger.error("Commission type is not pickleable")
         await self.setup_deadline_timer(commission)
```
This looks like it belongs in the `if set_success:` block as well.

zimmermatt reviewed (changes requested):
Looking good! Just one comment to take a look at.

Also, it will be helpful if you can point me at the commit with the text you'd like for the squashed merge that covers the entirety of the PR. Or, please write one up that I can use when you push this final change.

thompalex commented:
I think the following will make for a good commit message for the squash:

Subject: Created NotifyingServer and NotificationProtocol for our interaction with the Kademlia module

- NotificationServer and NotificationProtocol run the Kademlia module in the same way as usual, but they take in a callback function for the rpc_store method so that we can listen for and react to new items stored in the DHT. This will allow us to use the DHT as a sort of messaging system between peers.
- Adjusted the current peer class to await necessary asynchronous functions and modified our tests as needed to account for this change.
- Created a preliminary callback function to call on new store requests in our peer class. This callback will evolve to handle different types of "messages" or objects in the DHT as we progress with our goals.

zimmermatt commented on src/test/py/peer/peer_test.py:
```diff
@@ -86,21 +87,28 @@ async def test_commission_art_piece(self, mock_input):
+            self.test_logger.debug(mock_input)
+            self.peer.node = self.mock_node
+            await self.peer.commission_art_piece()
+            self.assertEqual(len(self.peer.commissions), 1)
```
At some point later, a test should be added that tests the failure case and asserts that the commissions count does not increase.

zimmermatt commented:
> I think the following will make for a good commit message for the squash:
> 
> Subject: Created NotifyingServer and NotificationProtocol for our interaction with the Kademlia module
> 
> * NotificationServer and NotificationProtocol run the Kademlia module in the same way as usual, but they take in a callback function for the rpc_store method so that we can listen for and react to new items stored in the DHT. This will allow us to use the DHT as a sort of messaging system between peers.
> * Adjusted the current peer class to await necessary asynchronous functions and modified our tests as needed to account for this change.
> * Created a preliminary callback function to call on new store requests in our peer class. This callback will evolve to handle different types of "messages" or objects in the DHT as we progress with our goals.

Looks good. I'm going to make a couple of minor revisions when I merge to comply with the style guide while maintaining the core of your version. Mainly:
* subject <= 50 chars
* imperative mood (guide says "in the subject line" but it also applies to body)
